docs/flex/interactive/configuration.md
# Configuration
## Configurable Items for Deploying Interactive with gsctl

When deploying Interactive using `gsctl`, various items can be configured. For a given configurable item named `item-name`, you can set its value as follows:

```bash
gsctl instance deploy --type interactive [--item-name=value]
```

Below is a list of all configurable items:

| Item Name         | Default | Description               | Since Version |
|-------------------|---------|---------------------------|---------------|
| coordinator-port  | 8080    | The port of the coordinator service  | v0.3          |
| admin-port       | 7777    | The port of the interactive admin service       | v0.3          |
| storedproc-port   | 10000    | The port of the interactive stored procedure service      | v0.3          |
| cypher-port       | 7687    | The port of the cypher service       | v0.3          |
| config | None    | The customized configuration file for engine interactive service | v0.4   |
| image-tag       | latest    | The version of the interactive you want to install | v0.5 |
| set             | None      | Specify additional properties for Interactive via the command line to override any corresponding settings in the configuration file if they are present | v0.5 |
<!-- | gremlin-port       | None    | The port of the gremlin service       | v0.3          | -->


<!-- *Note: The default value for `gremlin-port` is `None`, meaning the Gremlin service will not be initiated by default. -->

### Ports

By default, Interactive will launch the following services on these ports:

- Coordinator Service: 8080
- Interactive Meta Service: 7777
- Interactive Cypher Service: 7687
- Stored Procedure Service: 10000

You can customize these ports as needed. For example:

```bash
gsctl instance deploy --type interactive --coordinator-port 8081 --admin-port 7778 --cypher-port 7688 --storedproc-port 10001
```

<!-- ### Enabling Gremlin Service

The Gremlin service is disabled by default. To enable it, add the `--gremlin-port` option:

```bash
gsctl instance deploy --type interactive --coordinator-port 8081 --admin-port 7778 --cypher-port 7688 --storedproc-port 10001 --gremlin-port 8183
``` -->

### Service Configuration

By default, `Interactive` will initialize the service with its default settings.
However, GraphScope Interactive is designed to be flexible and adaptable to your specific needs. This means you can tailor the service's behavior using custom configurations.


#### Customizing Interactive Engine Service Configuration
To customize the service's settings, you can provide a YAML configuration file `interactive_config.yaml`. This file allows you to specify various parameters, from directory paths to log levels, ensuring the service aligns with your requirements. To use a custom configuration, simply pass the YAML file to the command as follows:

```bash
gsctl instance deploy --type interactive --config ./interactive_config.yaml
```

```{note}
Please be aware that you're not required to configure every option. Simply adjust the settings that are relevant to your needs. Any options left unconfigured will automatically adopt their default values, as detailed in the following sections.
```


##### Sample Configuration
Here's a glimpse of what a typical YAML configuration file might look like:

```yaml
log_level: INFO # default INFO, available(INFO,WARNING,ERROR,FATAL)
verbose_level: 0 # default 0, should be a int in range [0,10]. 10 will verbose all logs
compute_engine:
  thread_num_per_worker: 1  # the number of threads for each worker, default 1
compiler:
  planner:
    is_on: true
    opt: RBO
    rules:
      - FilterMatchRule
      - FilterIntoJoinRule
      - NotExistToAntiJoinRule
  query_timeout: 20000  # query timeout in milliseconds, default 20000
```

#### Sharded Service

The core query engine of Interactive is developed using [hiactor](https://github.com/alibaba/hiactor) which is based on [Seastar](https://github.com/scylladb/seastar). Seastar operates on a Share-nothing SMP architecture, where each core functions autonomously, without sharing memory, data structures, or CPU resources. Each Seastar core is commonly referred to as a shard.

Leveraging the future-promise API and a Cooperative micro-task scheduler, the sharded service significantly boosts performance and throughput. However, this setup can also lead to potential issues: an incoming request might experience delays even if some shards are idle, due to the shard scheduling algorithm potentially routing it to a busy shard. This can be problematic in Interactive, which typically hosts two services—`QueryService` and `AdminService`. Crucially, `AdminService` must remain responsive even when `QueryService` is under heavy load.

As discussed in [discussion-4409](https://github.com/alibaba/GraphScope/discussions/4409), one potential solution is to allocate different shards for handling distinct requests. This approach presents three scenarios:

- **Routine Scenario**: Here, users may execute both complex and simple queries, thus dedicating a shard exclusively for admin requests. However, since this shard won’t process queries, overall system performance may decline.
  
- **Performance-Critical Scenario**: In this scenario, users aim for peak performance from Interactive. All shards are used to process query requests, with admin requests being handled concurrently by them. Consequently, there may be instances of request delays.

By default, Interactive is configured for routine with the following:

```yaml
http_service:
  sharding_mode: exclusive # In exclusive mode, a shard is exclusively reserved for admin requests. In cooperative mode, both query and admin requests can be processed by any shard.
```

By changing to `sharding_mode: cooperative`, you can fully utilize all the computational power for the QueryService.


##### Available Configurations

In this following table, we use the `.` notation to represent the hierarchy within the `YAML` structure.


| PropertyName       | Default   | Meaning |  Since Version |
| --------           | --------  | -------- |-----------  |
| log_level     |  INFO   | The level of database log, INFO/WARNING/ERROR/FATAL | 0.0.1 |
| verbose_level     |  0   | The verbose level of database log, should be a int | 0.0.3 |
| compute_engine.thread_num_per_worker | 4 | The number of threads will be used to process the queries. Increase the number can benefit the query throughput | 0.0.1 |
| compute_engine.wal_uri    | file://{GRAPH_DATA_DIR}/wal | The location where Interactive will store and access WALs. `GRAPH_DATA_DIR` is a placeholder that will be populated by Interactive. | 0.5 |
| compute_engine.metadata_store.uri | file://{WORKSPACE}/METADATA | The location where Interactive will store and access metadatas. `WORKSPACE` is a placeholder that will be populated by Interactive. | 0.5 | 
| compiler.planner.is_on | true | Determines if query optimization is enabled for compiling Cypher queries  | 0.0.1 |
| compiler.planner.opt | RBO | Specifies the optimizer to be used for query optimization. Currently, only the Rule-Based Optimizer (RBO) is supported | 0.0.1 |
| compiler.planner.rules.FilterMatchRule | N/A | An optimization rule that pushes filter (`Where`) conditions into the `Match` clause | 0.0.1 |
| compiler.planner.rules.FilterIntoJoinRule | N/A | A native Calcite optimization rule that pushes filter conditions to the Join participants before performing the join | 0.0.1 |
| compiler.planner.rules.NotMatchToAntiJoinRule | N/A | An optimization rule that transforms a "not exist" pattern into an anti-join operation  | 0.0.1 |
| compiler.query_timeout  | 3000000   ｜ The maximum time for compiler to wait engine's reply, in `ms`  | 0.0.3 | 
| http_service.sharding_mode | exclusive | The sharding mode for http service, In exclusive mode, one shard is reserved exclusively for service admin request. In cooperative, both query request and admin request could be served by any shard. | 0.5 |
| http_service.max_content_length | 1GB | The maximum length of a http request that admin http service could handle | 0.5 |
| storage.string_default_max_length | 256 | The default maximum size for a string field | 0.5 |


#### TODOs

We currently only allow service configuration during instance deployment. In the near future, we will support:

- Graph-level configurations
- Modifying service configurations

docs/interactive_engine/interactive_db/configuration.md
# Configuration

To customize the settings of GraphScope Interactive, adjust the `gs_interactive.yaml` file. Please be aware that any changes to the configurations will only take effect after the service is [restarted](./getting_started.md).


## Available Configurations
For configurations associated with the root directory, we do not accept relative paths to ensure consistency.

### Service configurations

In this following table, we use the `.` notation to represent the hierarchy within the `YAML` structure.


| PropertyName       | Default   | Meaning |  Since Version |
| --------           | --------  | -------- |-----------  |
| version            |  latest   | The version of product version     |    0.0.1     |
| directories.workspace | /home/graphscope/interactive | The workspace of interactive docker container | 0.0.1 |
| directories.subdirs.data | data | The directory to store the graph indices and relate staffs | 0.0.1 |
| directories.subdirs.conf | conf | The directory to store the configuration of the Interactive database. | 0.0.1 |
| directories.subdirs.logs | logs | The directory to store the running log of database. | 0.0.1 |
| log_level     |  INFO   | The level of database log, INFO/DEBUG/ERROR | 0.0.1 |
|default_graph  | modern | The name of default graph on which to start the graph service. | 0.0.1 |
| compute_engine.type | hiactor | The backend query engine of databse, currenly only hiactor is supported. | 0.0.1 |
| compute_engine.hosts |localhost:10000 | The ip:port for the graph server to bind. | 0.0.1 |
| compute_engine.shard_num | 1 | The number of threads will be used to process the queries | 0.0.1 |
| compiler.planner | {"isOn":true,"opt":"RBO","rules":["FilterMatchRule"]} | The configuration of compiler planner | 0.0.1 |
| compiler.endpoint.bolt_connector.enabled | true | Turn on bolt service or not. | 0.0.1 |
| compiler.endpoint.bolt_connector.port | 7687 | The port bolt connector will listen on | 0.0.1 |
     

## A sample file of configurations
```yaml
---
version: 0.0.1
directories:
  workspace: /home/graphscope/interactive/
  subdirs:
    data: data  # by default data, relative to ${workspace}
    conf: conf # by default conf, relative to ${workspace}
    logs: logs  # by default logs, relative to ${workspace}
log_level: INFO # default INFO
default_graph: modern  # configure the graph to be loaded while starting the service, if graph name not specified
  # may include other configuration items of other engines
compute_engine:
  type: hiactor
  hosts:
    - localhost:10000  # currently only one host can be specified
  shared_num: 1  # the number of shared workers, default 1
compiler:
  planner: {"isOn":true,"opt":"RBO","rules":["FilterMatchRule"]} # Confirm这个配置
  endpoint:
    default_listen_address: localhost  # default localhost
    bolt_connector:   # for cypher, there may be other connectors, such as bolt_connector, https_connector
      enabled: true   # default false
      port: 7687
```



docs/release_notes.md
# Release Notes


## v0.18.0

This release contains many important features and enhancements to Graph Interactive Engine (GIE), including introducing a new strategy for pattern-matching queries and supporting canceling running queries. The Graph Learning Engine (GLE) now supports PyTorch and is also compatible with PyG. In addition, we take a first step towards modularized deployment for different components of GraphScope in Kubernetes.

We highlight the following improvements included in this release:

### Enhancements for GIE

A new execution strategy based on worst-case optimal join is introduced to GIE engine, potentially improving the performance of match step by orders of magnitude.
The query can be canceled after its execution exceeds a pre-given overtime parameter (10min by default).
GIE engine supports a failover mechanism: if an executor pod fails, but the data is not missing, it can be restarted by k8s. Existing queries cannot recover, but the engine can serve the following.

### Enhancements for GAE

Add more variants of WCC algorithm.
Supports local vertex map to make it could scale to larger graphs given more workers.

### Enhancements for GLE

Add support for PyTorch and PyG
Add heterogeneous graph support for subgraph-based GNN, add HeteroSubGraph and HeteroConv, bipartite GraphSAGE and UltraGCN.
Add edge feature support in both EgoGraph and SubGraph.
Add recommendation metrics: Recall, NDCG and HitRate.
Add hiactor-based graph engine.

### Standalone deployment

Supports a version of GIE standalone deployment in Kubernetes.
Redesigned docker files from the bottom to the top, making it more clear and more concise.
5. GAE Java Enhancement

Introduce @GrapeSkip for better user experience in template method overloading.
Speedup Java App runimte codegen.

Learn more about this release on [Release](https://github.com/alibaba/GraphScope/releases/tag/v0.18.0).

python/graphscope/config.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Copyright 2020 Alibaba Group Holding Limited. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

"""GraphScope default configuration."""

import base64
import json
from dataclasses import dataclass
from dataclasses import field
from typing import List
from typing import Union

from simple_parsing import ArgumentParser
from simple_parsing.helpers import Serializable
from simple_parsing.helpers import list_field

from graphscope.version import __version__

registry = "registry.cn-hongkong.aliyuncs.com"


@dataclass
class ResourceSpec:
    """Resource requirements for a container in kubernetes."""

    # CPU cores of container.
    cpu: Union[str, float, None] = None
    # Memory of container, suffix with ['Mi', 'Gi', 'Ti'].
    memory: Union[str, None] = None

    def as_dict(self):
        ret = {}
        if self.cpu is not None:
            ret["cpu"] = self.cpu
        if self.memory is not None:
            ret["memory"] = self.memory
        return ret


@dataclass
class ResourceConfig:
    """Resource spec for a container in kubernetes."""

    requests: ResourceSpec = None  # Resource requests of container.
    limits: ResourceSpec = None  # Resource limits of container.

    def get_requests(self):
        return self.requests.as_dict() if self.requests is not None else None

    def get_limits(self):
        return self.limits.as_dict() if self.limits is not None else None

    def set_cpu_request(self, cpu):
        self.requests.cpu = cpu
        # self.limits.cpu = cpu

    def set_mem_request(self, memory):
        self.requests.memory = memory
        # self.limits.memory = memory

    @staticmethod
    def make_burstable(cpu, memory):
        """Get default resource config for a container in kubernetes."""
        return ResourceConfig(
            requests=ResourceSpec(cpu=cpu, memory=memory),
            # limits=ResourceSpec(cpu=cpu, memory=memory),
        )


@dataclass
class ImageConfig:
    """Image related stuffs."""

    # k8s image registry.
    registry: Union[str, None] = "registry.cn-hongkong.aliyuncs.com"
    repository: str = "graphscope"  # k8s image repository.
    tag: str = __version__  # k8s image tag.
    # A list of secrets to pull image.
    pull_secrets: List[str] = field(default_factory=list)
    pull_policy: str = "IfNotPresent"  # Kubernetes image pull policy.


@dataclass
class MarsConfig:
    """Mars configuration"""

    enable: bool = False  # Enable Mars or not.
    worker_resource: ResourceConfig = field(
        default_factory=lambda: ResourceConfig.make_burstable(0.2, "4Mi")
    )
    scheduler_resource: ResourceConfig = field(
        default_factory=lambda: ResourceConfig.make_burstable(0.2, "4Mi")
    )


@dataclass
class DatasetConfig:
    """A Dataset container could be shipped with GraphScope in kubernetes."""

    enable: bool = False  # Mount the aliyun dataset bucket as a volume by ossfs.
    # A json string specifies the dataset proxy info. Available options of proxy: http_proxy, https_proxy, no_proxy.
    proxy: Union[str, None] = None


@dataclass
class EngineConfig:
    """Engine configuration"""

    enabled_engines: str = "gae,gie"  # A set of engines to enable.
    # Node selector for engine pods, default is None.
    node_selector: Union[str, None] = None

    enable_gae: bool = False  # Enable or disable analytical engine.
    # Enable or disable analytical engine with java support.
    enable_gae_java: bool = False
    enable_gie: bool = False  # Enable or disable interactive engine.
    enable_gle: bool = False  # Enable or disable graphlearn engine.
    enable_glt: bool = False  # Enable or disable graphlearn_torch engine.

    preemptive: bool = True

    # Resource for analytical pod
    gae_resource: ResourceConfig = field(
        default_factory=lambda: ResourceConfig.make_burstable(1, "4Gi")
    )

    # Resource for interactive executor pod
    gie_executor_resource: ResourceConfig = field(
        default_factory=lambda: ResourceConfig.make_burstable(1, "2Gi")
    )

    # Resource for interactive frontend pod
    gie_frontend_resource: ResourceConfig = field(
        default_factory=lambda: ResourceConfig.make_burstable(0.5, "1Gi")
    )

    # Resource for learning pod
    gle_resource: ResourceConfig = field(
        default_factory=lambda: ResourceConfig.make_burstable(0.2, "1Gi")
    )

    # Resource for learning pod
    glt_resource: ResourceConfig = field(
        default_factory=lambda: ResourceConfig.make_burstable(0.2, "1Gi")
    )

    def post_setup(self):
        valid_engines = set(
            "analytical,analytical-java,interactive,learning,gae,gae-java,gie,gle,glt".split(
                ","
            )
        )
        for item in [item.strip() for item in self.enabled_engines.split(",")]:
            if item not in valid_engines and item != "":
                print(f"Not a valid engine name: {item}")
            if item == "analytical" or item == "gae":
                self.enable_gae = True
            if item == "interactive" or item == "gie":
                self.enable_gie = True
            if item == "graphlearn" or item == "gle":
                self.enable_gle = True
            if item == "graphlearn-torch" or item == "glt":
                self.enable_glt = True
            if item == "analytical-java" or item == "gae-java":
                self.enable_gae_java = True

        if self.preemptive:
            self.gae_resource.requests = None
            self.gle_resource.requests = None
            self.glt_resource.requests = None
            self.gie_executor_resource.requests = None
            self.gie_frontend_resource.requests = None


@dataclass
class EtcdConfig:
    """Etcd configuration."""

    endpoint: Union[str, None] = None
    """The address of external etcd cluster, with formats like 'etcd01:port,etcd02:port,etcd03:port'.
    If address is set, all other etcd configurations are ignored.
    """

    # The port that etcd server will bind to for accepting client connections. Defaults to 2379.
    listening_client_port: int = 2379

    # The port that etcd server will bind to for accepting peer connections. Defaults to 2380.
    listening_peer_port: int = 2380

    # Kubernetes related config
    replicas: int = 1


@dataclass
class VineyardConfig:
    """Vineyard configuration"""

    # Vineyard IPC socket path, a socket suffixed by timestamp will be created in '/tmp' if not given.
    socket: Union[str, None] = None
    rpc_port: int = 9600  # Vineyard RPC port.

    # Kubernetes related config

    # The name of vineyard deployment, it should exist as expected.
    deployment_name: Union[str, None] = None

    image: str = "vineyardcloudnative/vineyardd:latest"  # Image for vineyard container.

    # Resource for vineyard sidecar container
    resource: ResourceConfig = field(
        default_factory=lambda: ResourceConfig.make_burstable(0.2, "256Mi")
    )


@dataclass
class InteractiveConfig:
    # a map from internal port to external port
    port_mapping: Union[dict, None] = None


@dataclass
class CoordinatorConfig:
    endpoint: Union[str, None] = None
    """The address of existed coordinator service, with formats like 'ip:port'.
    If address is set, all other coordinator configurations are ignored.
    """
    service_port: int = 63800  # Coordinator service port that will be listening on.

    http_port: int = 8080  # Coordinator HTTP service port

    monitor: bool = False  # Enable or disable prometheus exporter.
    monitor_port: int = 9090  # Coordinator prometheus exporter service port.

    # Kubernetes related config

    # Name of the coordinator deployment and service.
    deployment_name: Union[str, None] = None
    # Node selector for coordinator pod in kubernetes
    node_selector: Union[str, None] = None
    # Resource configuration of coordinator.
    resource: ResourceConfig = field(
        default_factory=lambda: ResourceConfig.make_burstable(0.5, "512Mi")
    )

    # For GraphScope operator
    # Launch coordinator only, do not let coordinator launch resources or delete resources.
    # It would try to find existing resources and connect to it.
    operator_mode: bool = False

    # For http server, limit the max content length of request. Mainly for file upload.
    max_content_length: str = "1G"

    # Only start coordinator http server, do not start others services.
    http_server_only: bool = False


@dataclass
class HostsLauncherConfig:
    """Local cluster configuration."""

    # list of hostnames of graphscope engine workers.
    hosts: List[str] = list_field("localhost")
    # Etcd configuration. Only local session needs to configure etcd.
    etcd: EtcdConfig = field(default_factory=EtcdConfig)

    # The number of retries when downloading dataset from internet.
    dataset_download_retries: int = 3


@dataclass
class KubernetesLauncherConfig:
    """Kubernetes cluster configuration."""

    # The namespace to create all resource, which must exist in advance.
    namespace: Union[str, None] = None
    delete_namespace: bool = False  # Delete the namespace that created by graphscope.

    config_file: Union[str, None] = None  # kube config file path

    # The deployment mode of engines on the kubernetes cluster, choose from 'eager' or 'lazy'.
    deployment_mode: str = "eager"

    # Service type, choose from 'NodePort' or 'LoadBalancer'.
    service_type: str = "NodePort"

    # A base64 encoded json string specifies the kubernetes volumes to mount.
    volumes: Union[str, None] = None

    # Wait until the graphscope instance has been deleted successfully.
    waiting_for_delete: bool = False

    image: ImageConfig = field(default_factory=ImageConfig)  # Image configuration.

    engine: EngineConfig = field(default_factory=EngineConfig)  # Engine configuration.

    # Dataset configuration.
    dataset: DatasetConfig = field(default_factory=DatasetConfig)

    mars: MarsConfig = field(default_factory=MarsConfig)  # Mars configuration.


@dataclass
class OperatorLauncherConfig:
    namespace: str = "default"
    gae_endpoint: str = ""
    hosts: List[str] = list_field()


@dataclass
class SessionConfig:
    """Session configuration"""

    num_workers: int = 2  # The number of graphscope engine workers.
    # The number of graphscope engine workers when launch local workers.
    default_local_num_workers: int = 1

    reconnect: bool = False  # Connect to an existed GraphScope Cluster
    instance_id: Union[str, None] = None  # Unique id for each GraphScope instance.

    # The length of time to wait before giving up launching graphscope.z
    timeout_seconds: int = 600
    # The length of time to wait starting from client disconnected before killing the graphscope instance.
    dangling_timeout_seconds: int = 600

    # The length of time to wait before retrying to launch graphscope.
    retry_time_seconds: int = 1

    execution_mode: str = "eager"  # The deploying mode of graphscope, eager or lazy.


@dataclass
class Config(Serializable):
    # Solution under the FLEX architecture, choose from 'GraphScope One', 'Interactive' or 'GraphScope Insight'
    solution: str = "GraphScope One"

    # Launcher type, choose from 'hosts', 'k8s' or 'operator'.
    launcher_type: str = "k8s"

    # Show log or not.
    show_log: bool = False
    # Log level, choose from 'info' or 'debug'.
    log_level: str = "info"

    session: SessionConfig = field(default_factory=SessionConfig)

    # Coordinator configuration.
    coordinator: CoordinatorConfig = field(default_factory=CoordinatorConfig)
    # Vineyard configuration.
    vineyard: VineyardConfig = field(default_factory=VineyardConfig)
    # Interactive configuration.
    interactive: InteractiveConfig = field(default_factory=InteractiveConfig)

    # Local cluster configuration.
    hosts_launcher: HostsLauncherConfig = field(default_factory=HostsLauncherConfig)

    # Kubernetes cluster configuration.
    kubernetes_launcher: KubernetesLauncherConfig = field(
        default_factory=KubernetesLauncherConfig
    )

    # Launcher used in operator mode.
    operator_launcher: OperatorLauncherConfig = field(
        default_factory=OperatorLauncherConfig
    )

    def set_option(self, key, value):  # noqa: C901
        """Forward set_option target to actual config fields"""
        if key == "addr":
            self.coordinator.endpoint = value
        elif key == "mode":
            self.session.execution_mode = value
        elif key == "cluster_type":
            self.launcher_type = value
        elif key == "k8s_namespace":
            self.kubernetes_launcher.namespace = value
        elif key == "k8s_image_registry":
            self.kubernetes_launcher.image.registry = value
        elif key == "k8s_image_repository":
            self.kubernetes_launcher.image.repository = value
        elif key == "k8s_image_tag":
            self.kubernetes_launcher.image.tag = value
        elif key == "k8s_image_pull_policy":
            self.kubernetes_launcher.image.pull_policy = value
        elif key == "k8s_image_secrets":
            self.kubernetes_launcher.image.pull_secrets = value
        elif key == "k8s_coordinator_cpu":
            self.coordinator.resource.set_cpu_request(value)
        elif key == "k8s_coordinator_mem":
            self.coordinator.resource.set_mem_request(value)
        elif key == "etcd_addrs":
            self.hosts_launcher.etcd.endpoint = value
        elif key == "etcd_listening_client_port":
            self.hosts_launcher.etcd.listening_client_port = value
        elif key == "etcd_listening_peer_port":
            self.hosts_launcher.etcd.listening_peer_port = value
        elif key == "k8s_vineyard_image":
            self.vineyard.image = value
        elif key == "k8s_vineyard_deployment":
            self.vineyard.deployment_name = value
        elif key == "k8s_vineyard_cpu":
            self.vineyard.resource.set_cpu_request(value)
        elif key == "k8s_vineyard_mem":
            self.vineyard.resource.set_mem_request(value)
        elif key == "k8s_engine_cpu":
            self.kubernetes_launcher.engine.gae_resource.set_cpu_request(value)
        elif key == "k8s_engine_mem":
            self.kubernetes_launcher.engine.gae_resource.set_mem_request(value)
        elif key == "mars_worker_cpu":
            self.kubernetes_launcher.mars.worker_resource.set_cpu_request(value)
        elif key == "mars_worker_mem":
            self.kubernetes_launcher.mars.worker_resource.set_mem_request(value)
        elif key == "mars_scheduler_cpu":
            self.kubernetes_launcher.mars.scheduler_resource.set_cpu_request(value)
        elif key == "mars_scheduler_mem":
            self.kubernetes_launcher.mars.scheduler_resource.set_mem_request(value)
        elif key == "k8s_coordinator_pod_node_selector":
            self.coordinator.node_selector = base64_encode(json.dumps(value))
        elif key == "k8s_engine_pod_node_selector":
            self.kubernetes_launcher.engine.node_selector = base64_encode(
                json.dumps(value)
            )
        elif key == "enabled_engines":
            self.kubernetes_launcher.engine.enabled_engines = value
        elif key == "with_mars":
            self.kubernetes_launcher.mars.enable = value
        elif key == "with_dataset":
            self.kubernetes_launcher.dataset.enable = value
        elif key == "k8s_volumes":
            self.kubernetes_launcher.volumes = base64_encode(json.dumps(value))
        elif key == "k8s_service_type":
            self.kubernetes_launcher.service_type = value
        elif key == "preemptive":
            self.kubernetes_launcher.engine.preemptive = value
        elif key == "k8s_deploy_mode":
            self.kubernetes_launcher.deployment_mode = value
        elif key == "k8s_waiting_for_delete":
            self.kubernetes_launcher.waiting_for_delete = value
        elif key == "num_workers":
            self.session.num_workers = value
            self.session.default_local_num_workers = value
        elif key == "show_log":
            self.show_log = value
        elif key == "log_level":
            self.log_level = value
        elif key == "timeout_seconds":
            self.session.timeout_seconds = value
        elif key == "dangling_timeout_seconds":
            self.session.dangling_timeout_seconds = value
        elif key == "dataset_download_retries":
            self.hosts_launcher.dataset_download_retries = value
        elif key == "k8s_client_config":
            self.kubernetes_launcher.config_file = value
        elif key == "reconnect":
            self.session.reconnect = value
        elif key == "vineyard_shared_mem":
            pass
        else:
            raise ValueError("Key not recognized: " + key)


def base64_encode(string):
    return base64.b64encode(string.encode("utf-8")).decode("utf-8", errors="ignore")


gs_config = Config()


if __name__ == "__main__":
    config = Config()
    config.coordinator.resource.requests = None
    config.kubernetes_launcher.image.registry = ""
    print(config.dumps_yaml())
    # print(config.dumps_json())
    s = config.dumps_yaml()
    config3 = Config.loads_yaml(s)
    print(config3)

    parser = ArgumentParser()
    parser.add_arguments(Config, dest="gs")
    args = parser.parse_args()
